src/lib/dnssd/IncrementalResolve.cpp

```diff
@@ -0,0 +1,374 @@
+
+    {
+        // TODO: Chip code historically seems to assume that the host name is of the
+        // form "<MAC or 802.15.4 Extended Address in hex>.local" and only saves the first part.
+        //
+        // This should not be needed as server name should not be relevant once parsed.
+        SerializedQNameIterator serverName = srv.GetName();
+
+        VerifyOrReturnError(serverName.Next() && serverName.IsValid(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        // only save the first part: the MAC or 802.15.4 Extended Address in hex
+        Platform::CopyString(mCommonResolutionData.hostName, serverName.Value());
+    }
+
+    switch (ComputeServiceNameType(name))
+    {
+    case ServiceNameType::kOperational:
+        mSpecificResolutionData.Set<OperationalNodeData>();
+        {
+            // Operational addresses start with peer node information
+            SerializedQNameIterator nameCopy = name;
+            if (!nameCopy.Next() || !nameCopy.IsValid())
+            {
+                mSpecificResolutionData = SpecificParseData();
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+
+            CHIP_ERROR err =
+                ExtractIdFromInstanceName(nameCopy.Value(), &mSpecificResolutionData.Get<OperationalNodeData>().peerId);
+            if (err != CHIP_NO_ERROR)
+            {
+                mSpecificResolutionData = SpecificParseData();
+                return err;
+            }
+        }
+        break;
+    case ServiceNameType::kCommissioner:
+    case ServiceNameType::kCommissionable:
+        mSpecificResolutionData.Set<CommissionNodeData>();
+        break;
+    default:
+        mSpecificResolutionData = SpecificParseData();
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+IncrementalResolver::RequiredInformationFlags IncrementalResolver::GetRequiredInformation() const
+{
+    RequiredInformationFlags flags;
+
+    if (!mSpecificResolutionData.Valid())
+    {
+        flags.Set(RequiredInformation::kSrvInitialization);
+    }
+    else
+    {
+        if (mCommonResolutionData.numIPs == 0)
+        {
+            flags.Set(RequiredInformation::kIpAddress);
+        }
+    }
+
+    return flags;
+}
+
+CHIP_ERROR IncrementalResolver::OnRecord(const ResourceData & data, BytesRange packetRange)
+{
+    if (!IsActive())
+    {
+        return CHIP_NO_ERROR; // nothing to parse
+    }
+
+    switch (data.GetType())
+    {
+    case QType::PTR:
+        return OnPtrRecord(data, packetRange);
+    case QType::TXT:
+        if (data.GetName() != mRecordName.Get())
+        {
+            ChipLogDetail(Discovery, "TXT record received for a different host name.");
+            return CHIP_NO_ERROR;
+        }
+        return OnTxtRecord(data, packetRange);
+    case QType::A: {
+        if (data.GetName() != mServerName.Get())
+        {
+            ChipLogDetail(Discovery, "IP address received for a different host name.");
```

Comment:
Suggest just adding tracing macros here, rather than logs that may massively pollute the logs record in case of bad records